<div style="border:solid Chocolate 2px; padding: 40px">

<b> Павел, привет!👋</b>

Меня зовут Кирилл Васильев, я буду ревьюером твоего проекта. Я предлагаю общаться на «ты», но если привычнее на «вы», просто скажи об этом! 

Я буду оставлять комментарии в твоем проекте. Пожалуйста, не удаляй их, они могут быть полезны в случае повторной проверки.

Ты можешь реагировать на комментарии как просто написав их в Markdown-ячейках, так и выделив их цветом. Например, <font color='blue'>синим</font>. Второй способ, на мой взгляд, удобнее.

Свои комментарии я буду обозначать <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным. Здесь же я могу давать советы и предложения.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самому найти и исправить ее. На реальной работе твой руководитель будет поступать также. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку!

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично. Есть описание проекта. Ясны цели и задачи.
</div>

## Подготовка данных

In [49]:
# код ревьюера - это нам нужно, чтобы избежать ошибки в OHE и такиудалить одну из колонок, чтобы избежать мультиколлинерарности
!pip install scikit-learn==1.1.3
!pip install category_encoders

In [50]:
!pip install imblearn

In [51]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [52]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score
from sklearn.utils import shuffle
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
from sklearn.metrics import confusion_matrix

In [1]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler

In [54]:
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler

In [55]:
from joblib import dump
from joblib import load

In [56]:
data = pd.read_csv('/datasets/Churn.csv')

Познакомимся с данными и проверим на мультиколлинеарность

In [57]:
data.shape

(10000, 14)

In [58]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [59]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [60]:
data.describe()

RowNumber    CustomerId   CreditScore           Age       Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  9091.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800     4.997690   
std     2886.89568  7.193619e+04     96.653299     10.487806     2.894723   
min        1.00000  1.556570e+07    350.000000     18.000000     0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000     2.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000     5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000     7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000    10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично, что смотрим на данные с помощью `describe()`, можно посмотреть также на переменные с нечисленными типами данных, если использовать аргумент `describe(include='all')`. Возможно, для восприятия будет удобно транспонировать таблицу.
</div>


<div class="alert alert-info"> 
Благодарю, ниже применю
</div>

In [61]:
data.describe(include='all')

RowNumber    CustomerId Surname   CreditScore Geography Gender  \
count   10000.00000  1.000000e+04   10000  10000.000000     10000  10000   
unique          NaN           NaN    2932           NaN         3      2   
top             NaN           NaN   Smith           NaN    France   Male   
freq            NaN           NaN      32           NaN      5014   5457   
mean     5000.50000  1.569094e+07     NaN    650.528800       NaN    NaN   
std      2886.89568  7.193619e+04     NaN     96.653299       NaN    NaN   
min         1.00000  1.556570e+07     NaN    350.000000       NaN    NaN   
25%      2500.75000  1.562853e+07     NaN    584.000000       NaN    NaN   
50%      5000.50000  1.569074e+07     NaN    652.000000       NaN    NaN   
75%      7500.25000  1.575323e+07     NaN    718.000000       NaN    NaN   
max     10000.00000  1.581569e+07     NaN    850.000000       NaN    NaN   

                 Age       Tenure        Balance  NumOfProducts    HasCrCard  \
count   10000.000000  9091.000000   10000.000000   10000.000000  10000.00000   
unique           NaN          NaN            NaN            NaN          NaN   
top              NaN          NaN            NaN            NaN          NaN   
freq             NaN          NaN            NaN            NaN          NaN   
mean       38.921800     4.997690   76485.889288       1.530200      0.70550   
std        10.487806     2.894723   62397.405202       0.581654      0.45584   
min        18.000000     0.000000       0.000000       1.000000      0.00000   
25%        32.000000     2.000000       0.000000       1.000000      0.00000   
50%        37.000000     5.000000   97198.540000       1.000000      1.00000   
75%        44.000000     7.000000  127644.240000       2.000000      1.00000   
max        92.000000    10.000000  250898.090000       4.000000      1.00000   

        IsActiveMember  EstimatedSalary        Exited  
count     10000.000000     10000.000000  10000.000000  
unique             NaN              NaN           NaN  
top                NaN              NaN           NaN  
freq               NaN              NaN           NaN  
mean          0.515100    100090.239881      0.203700  
std           0.499797     57510.492818      0.402769  
min           0.000000        11.580000      0.000000  
25%           0.000000     51002.110000      0.000000  
50%           1.000000    100193.915000      0.000000  
75%           1.000000    149388.247500      0.000000  
max           1.000000    199992.480000      1.000000

In [62]:
data.corr()

RowNumber  CustomerId  CreditScore       Age    Tenure  \
RowNumber         1.000000    0.004202     0.005840  0.000783 -0.007322   
CustomerId        0.004202    1.000000     0.005308  0.009497 -0.021418   
CreditScore       0.005840    0.005308     1.000000 -0.003965 -0.000062   
Age               0.000783    0.009497    -0.003965  1.000000 -0.013134   
Tenure           -0.007322   -0.021418    -0.000062 -0.013134  1.000000   
Balance          -0.009067   -0.012419     0.006268  0.028308 -0.007911   
NumOfProducts     0.007246    0.016972     0.012238 -0.030680  0.011979   
HasCrCard         0.000599   -0.014025    -0.005458 -0.011721  0.027232   
IsActiveMember    0.012044    0.001665     0.025651  0.085472 -0.032178   
EstimatedSalary  -0.005988    0.015271    -0.001384 -0.007201  0.010520   
Exited           -0.016571   -0.006248    -0.027094  0.285323 -0.016761   

                  Balance  NumOfProducts  HasCrCard  IsActiveMember  \
RowNumber       -0.009067       0.007246   0.000599        0.012044   
CustomerId      -0.012419       0.016972  -0.014025        0.001665   
CreditScore      0.006268       0.012238  -0.005458        0.025651   
Age              0.028308      -0.030680  -0.011721        0.085472   
Tenure          -0.007911       0.011979   0.027232       -0.032178   
Balance          1.000000      -0.304180  -0.014858       -0.010084   
NumOfProducts   -0.304180       1.000000   0.003183        0.009612   
HasCrCard       -0.014858       0.003183   1.000000       -0.011866   
IsActiveMember  -0.010084       0.009612  -0.011866        1.000000   
EstimatedSalary  0.012797       0.014204  -0.009933       -0.011421   
Exited           0.118533      -0.047820  -0.007138       -0.156128   

                 EstimatedSalary    Exited  
RowNumber              -0.005988 -0.016571  
CustomerId              0.015271 -0.006248  
CreditScore            -0.001384 -0.027094  
Age                    -0.007201  0.285323  
Tenure                  0.010520 -0.016761  
Balance                 0.012797  0.118533  
NumOfProducts           0.014204 -0.047820  
HasCrCard              -0.009933 -0.007138  
IsActiveMember         -0.011421 -0.156128  
EstimatedSalary         1.000000  0.012097  
Exited                  0.012097  1.000000

Мультколлинеарности нет, но есть не нужные ждя модели ненужные столбцы, например RowNumber, CustomerId и Surname. Эти данные быди бы полезны дня анализа жанных, но для модели бесполезны, и даже скорее всего вредны. Удалим столбцы. 

In [63]:
data = data.drop(['RowNumber', 'CustomerId', 'Surname'], axis=1)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Согласен. Отличное обоснование!
</div>


In [64]:
data.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42     2.0       0.00              1   
1          608     Spain  Female   41     1.0   83807.86              1   
2          502    France  Female   42     8.0  159660.80              3   
3          699    France  Female   39     1.0       0.00              2   
4          850     Spain  Female   43     2.0  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

Есть пропуски в столбце Tenure. Изучим столбец и посмотрим что можно сделать

In [65]:
data['Tenure'].value_counts()

1.0     952
2.0     950
8.0     933
3.0     928
5.0     927
7.0     925
4.0     885
9.0     882
6.0     881
10.0    446
0.0     382
Name: Tenure, dtype: int64

Не читая 10 и 0, данные очень равноерно распределены. Думаю лучше удалить строки с Null, это меньше 10% данных.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Согласен. Это допустимое решение. 
</div>


In [66]:
data = data.dropna(subset=['Tenure']).reset_index(drop=True)

In [67]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9091 entries, 0 to 9090
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CreditScore      9091 non-null   int64  
 1   Geography        9091 non-null   object 
 2   Gender           9091 non-null   object 
 3   Age              9091 non-null   int64  
 4   Tenure           9091 non-null   float64
 5   Balance          9091 non-null   float64
 6   NumOfProducts    9091 non-null   int64  
 7   HasCrCard        9091 non-null   int64  
 8   IsActiveMember   9091 non-null   int64  
 9   EstimatedSalary  9091 non-null   float64
 10  Exited           9091 non-null   int64  
dtypes: float64(3), int64(6), object(2)
memory usage: 781.4+ KB


In [68]:
data.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42     2.0       0.00              1   
1          608     Spain  Female   41     1.0   83807.86              1   
2          502    France  Female   42     8.0  159660.80              3   
3          699    France  Female   39     1.0       0.00              2   
4          850     Spain  Female   43     2.0  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

Очень интересно получается. Страны у нас три и с одной стороны хочется применить технику Ordinal Encoding, но она очень плохо работает с логистической регрессией поэтому скорее всего остановимся на технике One-Hot Encoding и   скорее всего применим модель логистической регрессии поскольку такая модель очень хорошо подходит для решения задач классификации. В учебных целя также можно разоьрать модели на основе деревьев. Забегая вперёд отмечу что модели основанные на деревьях, а именно решающее дерево и случайный лес не будем использовать поскольку данные модели плохо реагируют на масштабирование данных, а судя по задаче данные масштабировать придётся.

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Да, действительно, регрессионные модели не работают с порядковым кодированием, «деревянные» же хорошо с ним справляются. Масштабирование же признаков тоже не должно негативно сказываться на деревьях.
    
Достаточно хорошо это объясняется здесь https://habr.com/ru/company/ods/blog/322534/ в пункте «Как дерево решений работает с количественными признаками».
    
PS Забегая вперед, скажу, что ниже ты попробовал модели деревьев до и после масштабирования и убедился, что результат идентичный. Так что, вероятно, нужно подправить комментарий выше :)
       
</div>


<div class="alert alert-info"> 
Благодарю, дабы не терять ход мысли изменять не буду вывод, но по ходу проекта буду добавлять новые выводы
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Ок, договорились. Ну я больше здесь тогда оставил комментарий на будущее - когда будем готовить проекты к резюме. Нужно будет причесать.
</div>

In [69]:
#data['Geography'].value_counts()

In [70]:
#data['Gender'].value_counts()

Применим One-Hot Encoding

In [71]:
#data['Geography'] = pd.get_dummies(data['Geography'], drop_first=True)
#data['Gender'] = pd.get_dummies(data['Gender'], drop_first=True)

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

В тренажере давался пример использования прямого кодирования с помощью get_dummies, однако это далеко не самый оптимальный вариант. Он подходит для анализа данных, но в машинном обучении может служить причиной ошибок при обучении моделей.
    
Предлагаю тебе использовать класс <a href="https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html">OHE из sklearn</a>, тем более, что мы импортируем его в начале проекта.
      
</div>

<div class="alert alert-info"> 
Благодарю, исправлю
</div>

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
И сразу уточню, что OHE мы используем уже после разделения выборок на обучающую, валидационную и тестовую. Обучаем на трейне, применяем ко всем.
    
В этом плане кодирование в нашем алгоритме работы будет иднетично с масштабированием: на том же этапе, также fit к трейну, а затем применяем на всех выборках.

</div>


<div class="alert alert-info"> 
Благодарю, исправлю
</div>

In [72]:
data.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42     2.0       0.00              1   
1          608     Spain  Female   41     1.0   83807.86              1   
2          502    France  Female   42     8.0  159660.80              3   
3          699    France  Female   39     1.0       0.00              2   
4          850     Spain  Female   43     2.0  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Дисбаланс целевого признака также лучше проверить до начала обучения моделей. Ты делаешь это после.
       
</div>




<div class="alert alert-info"> 
Благодарю, исправил
</div>

In [73]:
data['Exited'].value_counts()

0    7237
1    1854
Name: Exited, dtype: int64

<div class="alert alert-warning">
<h2> Комментарий ревьюера  #2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Ок, ну и предполагается, что в двух словах скажем, что не дело это иметь такой дисбаланс.
       
</div>


<div class="alert alert-info"> 
Благодарю, я просто это в тот раз делал. Но замечение правильное.
</div>

Наблюдается дисбаланс классов, то не хорошо. Ниже будем избавляться от этого явления

## Исследование задачи

Разобьем данные на выборки

In [74]:
target = data['Exited']
features = data.drop('Exited', axis=1)

features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.25, random_state=12345)
features_valid, features_test, target_valid, target_test = train_test_split(
    features_valid, target_valid, test_size=0.5, random_state=12345)

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Ок, по размерам выборок все хорошо. Было бы неплохо здесь вывести размер выборок `shape`.
       
</div>


<div class="alert alert-info"> 
Благодарю, исправил
</div>

In [75]:
print(data.shape)
print('-------------------')
print(features_train.shape)
print(target_train.shape)
print(features_valid.shape)
print(target_valid.shape)
print(features_test.shape)
print(target_test.shape)

(9091, 11)
-------------------
(6818, 10)
(6818,)
(1136, 10)
(1136,)
(1137, 10)
(1137,)


<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Да, отлично. Так удобнее, мне кажется, понимать, что везде данных хватает и по количеству, и по пропорциям. Хорошее разделение.
</div>
 

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

На этом этапе нужно будет закодировать категориальные признаки с помощью OHE.
       
</div>


<div class="alert alert-info"> 
Благодарю,   я потратил целый день, но так и не разоьрался как кодировать. Перепробывал все, что можно. Информации в сети практически не нашел. Везде выдает ошибку. Прошу жать подсказку 
</div>

<div class="alert alert-warning">
<h2> Комментарий ревьюера  #2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

На этом моменте после прочтения твоего комментария я начал с того, что загрузил версию энкодера в самом начале проекта (посмотри). 
    
`!pip install scikit-learn==1.1.3`
    
`!pip install category_encoders`
    
Есть такая история, что энкодер летает через раз при одновременном применении параметров (handle_unknown='ignore', drop='first'), а они то нам и нужны. Что они означают? Дропаем первую колонку - ну тут понятно, боремся с мультиколлинеарностью. handle_unknown = что делаем, если после обучения встретим неизвестную категорию (на трейн не попала, а в тесте вылезла, например). Параметр 'ignore' - игнорируем, не падаем, работаем дальше. Очень полезно.

    
Дальше ты называешь колонки numeric - а ведь у нас именно они то и не «numeric». Может, как-то переназвать, чтобы не путаться? 
    
Берем энкодер, обучаем на нужных колонках, дальше применяем ко всем трем выборкам (я применил к двум, за тобой тест).
    
Дальше я вывожу, что у нас вышло из энкодера. Обрати внимание, там только те колонки, которые мы закодировали. Нужно их будет объединить со всеми остальными. Обрати внимание, энкодер не выдает датафрейм, мы к нему приводим сами.
       
</div>

<div class="alert alert-info"> 
Благодарю, но этот код тоже выдает ошибку
</div>

<div class="alert alert-warning">
<h2> Комментарий ревьюера  #2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Ух )) Чтобы изменения в импортах применились, нужно перезагрузить среду исполнения: вверху на меню kernel - Restart&Run All.
       
</div>


<div class="alert alert-info"> 
Благодарю, нам это просто не рассказывали
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Да оно вроде бы недавно падать начало все с OHE ))) Но все равно нужно сейчас освоить этот метод, потому что в дальнейших проектах он пригодится и там будет много чего другого интересного.
    
А про «перезагрузить» - тут да, много было бы чуть поподробнее рассказать, согласен.
</div>

In [76]:
categoricals = ['Geography', 'Gender']

"""

ohe = OneHotEncoder(categories = 'auto', sparse = False, drop='first')
ohe.fit(features_train[['Geography']])

features_train[['Geography']] = ohe.transform(features_train[['Geography']]).toarray()


"""

#код ревьюера

encoder = OneHotEncoder(handle_unknown='ignore', drop='first') 

features_train_ohe = pd.DataFrame(
    encoder.fit_transform(features_train[categoricals]).toarray(),
    columns=encoder.get_feature_names_out()
)

features_valid_ohe = pd.DataFrame(
    encoder.transform(features_valid[categoricals]).toarray(),
    columns=encoder.get_feature_names_out()
)

# дальше тест нужно будет прогнать еще не забыть (все по аналогии с валидационной)

features_test_ohe = pd.DataFrame(
    encoder.transform(features_test[categoricals]).toarray(),
    columns=encoder.get_feature_names_out()
)

In [77]:
# код ревьюера
features_train_ohe.head()

Geography_Germany  Geography_Spain  Gender_Male
0                1.0              0.0          0.0
1                0.0              0.0          1.0
2                1.0              0.0          1.0
3                0.0              1.0          1.0
4                0.0              0.0          1.0

<div class="alert alert-warning">
<h2> Комментарий ревьюера  #2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

На этом месте я бы отвлекся на масштабирование количественных признаков (нумерик как раз), а затем уже объединял бы две части в готовый датафрейм.
       
</div>

<div class="alert alert-info"> 
Благодарю, исправил
</div>

У нас есть признаки CreditScore, Balance, EstimatedSalary и их значения сильно разьроссаны, поэтому алгоритм будет думать, что какойто признак важнее другого. Приведем признаки к одному масштабу.

In [78]:
pd.options.mode.chained_assignment = None

numeric = ['CreditScore', 'Balance', 'EstimatedSalary']

scaler = StandardScaler()
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric]) 

In [79]:
features_train[numeric]

CreditScore   Balance  EstimatedSalary
1349    -1.037187  0.937205        -1.635163
6740     0.297802 -1.233158        -0.590123
6816    -1.078582  0.307627         0.073185
5145    -1.130326 -1.233158         1.674643
6678    -0.478354  0.420246        -1.069473
...           ...       ...              ...
4478     1.280934  0.808482        -1.104143
4094     0.152920  1.065897         0.254945
3492    -1.057885  0.446252        -1.098318
2177    -0.964746 -1.233158         1.286788
4578     1.218841  1.010477        -1.255989

[6818 rows x 3 columns]

In [80]:
# код ревьюера
# здесь главное не забыть объединить.


#features_train_enc = features_train_ohe.join(features_train_normalised)

# то же самое с валидацией и тестом

In [81]:
features_train

CreditScore Geography  Gender  Age  Tenure   Balance  NumOfProducts  \
1349    -1.037187   Germany  Female   43     3.0  0.937205              1   
6740     0.297802    France    Male   38     3.0 -1.233158              2   
6816    -1.078582   Germany    Male   37     9.0  0.307627              2   
5145    -1.130326     Spain    Male   28     8.0 -1.233158              2   
6678    -0.478354    France    Male   40     4.0  0.420246              2   
...           ...       ...     ...  ...     ...       ...            ...   
4478     1.280934   Germany    Male   34    10.0  0.808482              1   
4094     0.152920    France  Female   44     8.0  1.065897              1   
3492    -1.057885    France  Female   29     6.0  0.446252              1   
2177    -0.964746    France  Female   52     9.0 -1.233158              1   
4578     1.218841   Germany    Male   23     2.0  1.010477              1   

      HasCrCard  IsActiveMember  EstimatedSalary  
1349          1               0        -1.635163  
6740          1               0        -0.590123  
6816          0               1         0.073185  
5145          0               0         1.674643  
6678          1               1        -1.069473  
...         ...             ...              ...  
4478          0               0        -1.104143  
4094          1               1         0.254945  
3492          1               1        -1.098318  
2177          1               0         1.286788  
4578          0               1        -1.255989  

[6818 rows x 10 columns]

Удалим ячейки Geography и Gender перед обьединением. Они больше не нужны

In [82]:
features_train = features_train.drop(['Geography', 'Gender'], axis=1, errors='ignore')
features_valid = features_valid.drop(['Geography', 'Gender'], axis=1, errors='ignore')
features_test = features_test.drop(['Geography', 'Gender'], axis=1, errors='ignore')

<div class="alert alert-info"> 
Здесь приглось применить errors='ignore' поскольку без ьыла ошибка label not contained in axis
</div>

In [83]:
features_train.head()

CreditScore  Age  Tenure   Balance  NumOfProducts  HasCrCard  \
1349    -1.037187   43     3.0  0.937205              1          1   
6740     0.297802   38     3.0 -1.233158              2          1   
6816    -1.078582   37     9.0  0.307627              2          0   
5145    -1.130326   28     8.0 -1.233158              2          0   
6678    -0.478354   40     4.0  0.420246              2          1   

      IsActiveMember  EstimatedSalary  
1349               0        -1.635163  
6740               0        -0.590123  
6816               1         0.073185  
5145               0         1.674643  
6678               1        -1.069473

In [84]:
#код ревьюера
features_train= features_train.reset_index(drop=True)
features_valid = features_valid.reset_index(drop=True)
features_test = features_test.reset_index(drop=True)

<div class="alert alert-info"> 
Благодарю, забегая вперед скажу, что отсутствие 0 и правильные индексы сделали свое дело. Ниже модели дают f1-0.85 и auc-0.9
</div>

In [117]:
features_train_ohe.head()

Geography_Germany  Geography_Spain  Gender_Male
0                1.0              0.0          0.0
1                0.0              0.0          1.0
2                1.0              0.0          1.0
3                0.0              1.0          1.0
4                0.0              0.0          1.0

Обьединим данные

In [86]:
features_train_enc = features_train_ohe.join(features_train)
features_valid_enc = features_valid_ohe.join(features_valid)
features_test_enc = features_test_ohe.join(features_test)

In [87]:
#features_train_enc = features_train_ohe.merge(features_train, how='right', left_index=True, right_index=True)
#features_valid_enc = features_valid_ohe.merge(features_valid, how='right', left_index=True, right_index=True)
#features_test_enc = features_test_ohe.merge(features_test, how='right', left_index=True, right_index=True)

In [88]:
features_train_enc

Geography_Germany  Geography_Spain  Gender_Male  CreditScore  Age  \
0                   1.0              0.0          0.0    -1.037187   43   
1                   0.0              0.0          1.0     0.297802   38   
2                   1.0              0.0          1.0    -1.078582   37   
3                   0.0              1.0          1.0    -1.130326   28   
4                   0.0              0.0          1.0    -0.478354   40   
...                 ...              ...          ...          ...  ...   
6813                1.0              0.0          1.0     1.280934   34   
6814                0.0              0.0          0.0     0.152920   44   
6815                0.0              0.0          0.0    -1.057885   29   
6816                0.0              0.0          0.0    -0.964746   52   
6817                1.0              0.0          1.0     1.218841   23   

      Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0        3.0  0.937205              1          1               0   
1        3.0 -1.233158              2          1               0   
2        9.0  0.307627              2          0               1   
3        8.0 -1.233158              2          0               0   
4        4.0  0.420246              2          1               1   
...      ...       ...            ...        ...             ...   
6813    10.0  0.808482              1          0               0   
6814     8.0  1.065897              1          1               1   
6815     6.0  0.446252              1          1               1   
6816     9.0 -1.233158              1          1               0   
6817     2.0  1.010477              1          0               1   

      EstimatedSalary  
0           -1.635163  
1           -0.590123  
2            0.073185  
3            1.674643  
4           -1.069473  
...               ...  
6813        -1.104143  
6814         0.254945  
6815        -1.098318  
6816         1.286788  
6817        -1.255989  

[6818 rows x 11 columns]

<div class="alert alert-warning">
<h2> Комментарий ревьюера  #2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Я бы попробовал удалить строки с пропуском, их, конечно, много, но интересно, улучшится ли качество.
    
Откуда взялись нули − вопрос. Нужно бы проверить качество склейки. 
       
</div>

<div class="alert alert-info"> 
Трудно сказать правильно я сделал что поставил везде 0 или лучше ьыло звнулить. Еше можно было обьединить по merge(features_train, how='right', left_index=True, right_index=True). Nan были бы в признаках к которым применили OHE
</div>

<div class="alert alert-warning">
<h2> Комментарий ревьюера  #2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Да вроде бы пропуски как раз не в них. Если в них, то логично занулить, да.
       
</div>

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
 
И здесь же нужно будет масштабировать количественные признаки, если дальше мы используем регрессионные модели. Ты в проекте делаешь это ниже, но оптимальнее это сделать уже на этом этапе, чтобы не запускать модель заведомо без шансов на хороший результат.

</div>


<div class="alert alert-info"> 
Благодарю, все что сделал в первый раз закоментарю тогда
</div>

Построим модель дерева решений

In [89]:
for depth in range(1, 50):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
    model.fit(features_train_enc, target_train) 
    predicted_valid = model.predict(features_valid_enc)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_valid_enc)
    probabilities_one_valid = probabilities_valid[:, 1]

    fpr, tpr, treshold = roc_curve(target_valid, probabilities_one_valid)
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
    print('depth:', depth, ' ', 'F1:', f1, ' ', 'auc_roc:', auc_roc)

depth: 1   F1: 0.0   auc_roc: 0.676547653325655
depth: 2   F1: 0.5643564356435643   auc_roc: 0.7436030329206257
depth: 3   F1: 0.44012944983818775   auc_roc: 0.7920001919570017
depth: 4   F1: 0.5730027548209367   auc_roc: 0.8235939149630483
depth: 5   F1: 0.525679758308157   auc_roc: 0.8498272386985316
depth: 6   F1: 0.5714285714285714   auc_roc: 0.8499496112870717
depth: 7   F1: 0.5691906005221932   auc_roc: 0.844217295325847
depth: 8   F1: 0.547872340425532   auc_roc: 0.812517995968903
depth: 9   F1: 0.5241730279898219   auc_roc: 0.7782896631154621
depth: 10   F1: 0.5190839694656488   auc_roc: 0.7660763988866494
depth: 11   F1: 0.5036674816625917   auc_roc: 0.7265308570880122
depth: 12   F1: 0.4834123222748815   auc_roc: 0.7083741241961801
depth: 13   F1: 0.5069124423963134   auc_roc: 0.6926024570496208
depth: 14   F1: 0.49448123620309054   auc_roc: 0.6937374028217679
depth: 15   F1: 0.4736842105263158   auc_roc: 0.6653613590555716
depth: 16   F1: 0.4592274678111588   auc_roc: 0.6597

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Здесь нам важно также вывести, при каких гиперпараметрах достигается такой результат. То есть, нужно подавать на принт не только метрику, но и соответствующую ей глубину дерева.
    
(предположу, что от этой глубины можно будет оттолкнуться, когда в случайном лесе будем перебирать гиперпараметры).

Также напомню, что по ТЗ проекта нам нужно измерить и ROC-AUC: *«Дополнительно измеряйте AUC-ROC, сравнивайте её значение с F1-мерой».*
</div>


<div class="alert alert-info"> 
Благодарю, исправил
</div>

In [90]:
#Видно что на значении delf = 7 достигается наилучший результат по обучающей выборке 0.5729

Наилучший результат при depth: 31   F1: 0.05303030303030304   auc_roc: 0.603992705633938. f1 мера очень низкая, попробуем другую модель

Построим модель случайного леса

In [91]:
for est in range(1, 200, 20):
    for depth in range(40, 100, 10):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth) 
        model.fit(features_train_enc, target_train) 
        predicted_valid = model.predict(features_valid_enc)
        f1 = f1_score(target_valid, predicted_valid)
        probabilities_valid = model.predict_proba(features_valid_enc)
        probabilities_one_valid = probabilities_valid[:, 1]

        fpr, tpr, treshold = roc_curve(target_valid, probabilities_one_valid)
        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
        print('depth:', depth, ' ', 'F1:', f1, ' ', 'est:', est,  ' ', 'auc_roc:', auc_roc)

depth: 40   F1: 0.4618556701030928   est: 1   auc_roc: 0.6645599385737595
depth: 50   F1: 0.4618556701030928   est: 1   auc_roc: 0.6645599385737595
depth: 60   F1: 0.4618556701030928   est: 1   auc_roc: 0.6645599385737595
depth: 70   F1: 0.4618556701030928   est: 1   auc_roc: 0.6645599385737595
depth: 80   F1: 0.4618556701030928   est: 1   auc_roc: 0.6645599385737595
depth: 90   F1: 0.4618556701030928   est: 1   auc_roc: 0.6645599385737595
depth: 40   F1: 0.5818181818181818   est: 21   auc_roc: 0.8443396679143871
depth: 50   F1: 0.5818181818181818   est: 21   auc_roc: 0.8443396679143871
depth: 60   F1: 0.5818181818181818   est: 21   auc_roc: 0.8443396679143871
depth: 70   F1: 0.5818181818181818   est: 21   auc_roc: 0.8443396679143871
depth: 80   F1: 0.5818181818181818   est: 21   auc_roc: 0.8443396679143871
depth: 90   F1: 0.5818181818181818   est: 21   auc_roc: 0.8443396679143871
depth: 40   F1: 0.5851063829787234   est: 41   auc_roc: 0.8455322007870236
depth: 50   F1: 0.5851063829787

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Я бы здесь применил двойной цикл. Можно увеличить шаг в количестве оценщиков `n_estimators`, но попробовать несколько вариантов с глубиной дерева.
       
</div>


<div class="alert alert-info"> 
Благодарю, сделал
</div>

In [92]:
#Видно что результат не удучшился, лучший на est=200, 0.56980

Наилучший результат при depth: 40-90, F1: 0.048780487804878044   est 21   auc_roc: 0.6466551492465689, f1 мера очень низкая, попробуем другую модель

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Здесь то же замечание − нужно выводить не только значение метрики качества, но и гиперпараметры. Пока неочевидно, откуда взяли `est=200`.

</div>


<div class="alert alert-info"> 
Благодарю, исправил
</div>

Построим модель логистической регрессии 

In [93]:
model = LogisticRegression(random_state=12345, solver='liblinear')
model.fit(features_train_enc, target_train)
predicted_valid = model.predict(features_valid_enc)
f1 = f1_score(target_valid, predicted_valid)
probabilities_valid = model.predict_proba(features_valid_enc)
probabilities_one_valid = probabilities_valid[:, 1]
fpr, tpr, treshold = roc_curve(target_valid, probabilities_one_valid)
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
print('F1:', f1, ' ', 'auc_roc:', auc_roc)

F1: 0.3470031545741325   auc_roc: 0.7820472214224014


In [94]:
#Вопреки моей первой гипотезе логистическая рекгрессия показала плохой рузультат.
#У нас есть признаки CreditScore, Balance, EstimatedSalary  и их значения сильно разьроссаны, поэтому алгоритм будет думать, что какойто признак важнее другого. 
#Приведем признаки к одному масштабу.

Результат никакой от слова совсем.Возможна причина в дисбалансе классов.Проверим

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Верная догадка (выше написал, на каком моменте лучше сделать нормализацию).
</div>

In [95]:
#pd.options.mode.chained_assignment = None

#numeric = ['CreditScore', 'Balance', 'EstimatedSalary']

#scaler = StandardScaler()
#scaler.fit(features_train[numeric])
#features_train[numeric] = scaler.transform(features_train[numeric])
#features_valid[numeric] = scaler.transform(features_valid[numeric])
#features_test[numeric] = scaler.transform(features_test[numeric])

In [96]:
#Снова попробуем построить модели

In [97]:
#model = DecisionTreeClassifier(random_state=12345, max_depth=7) 
#model.fit(features_train, target_train) 
#predicted_valid = model.predict(features_valid)
#f1 = f1_score(target_valid, predicted_valid)
    
#print(f1)

In [98]:
#model = RandomForestClassifier(random_state=12345, n_estimators=200, max_depth=7) 
#model.fit(features_train, target_train) 
#predicted_valid = model.predict(features_valid)
#f1 = f1_score(target_valid, predicted_valid)
    
#print(f1)

In [99]:
#model = LogisticRegression(random_state=12345, solver='liblinear')
#model.fit(features_train, target_train)
#predicted_valid = model.predict(features_valid)
#print("f1_score:", f1_score(target_valid, predicted_valid)) 

In [100]:
#Маштабирование немного улучшило ситуацию c регрессией, и не поалияло на модели с деревьями, что соответствует теории. Но думаю иммет смысл проверки на дисболанс классов

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

А выше ты писал, что масштабирование как-то сказывается на деревьях :)
       
</div>


In [101]:
# data['Exited'].value_counts()

In [102]:
#Что и требовалось доказать, наблюдается дисбаланс класслов

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Ага.
</div>

## Борьба с дисбалансом

Для борьбы с дисбалансом классов применим несколько стрптегий.
1.  class_weight = 'balanced' добавим этот аргумент к каждой модели
2. upsampling
3. downsampling

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отличный план.
</div>

1 стратегия

Дерево решений

In [103]:
for depth in range(1, 50):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth, class_weight = 'balanced') 
    model.fit(features_train_enc, target_train) 
    predicted_valid = model.predict(features_valid_enc)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_valid_enc)
    probabilities_one_valid = probabilities_valid[:, 1]

    fpr, tpr, treshold = roc_curve(target_valid, probabilities_one_valid)
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
    print('depth:', depth, ' ', 'F1:', f1, ' ', 'auc_roc:', auc_roc)

depth: 1   F1: 0.4655737704918032   auc_roc: 0.6773490738074671
depth: 2   F1: 0.4944178628389156   auc_roc: 0.7437613974469718
depth: 3   F1: 0.5197934595524957   auc_roc: 0.7908652461848545
depth: 4   F1: 0.5307346326836582   auc_roc: 0.8130434782608695
depth: 5   F1: 0.5685279187817259   auc_roc: 0.8317136961320664
depth: 6   F1: 0.5533333333333333   auc_roc: 0.8275434302716191
depth: 7   F1: 0.5591766723842195   auc_roc: 0.8222262213264231
depth: 8   F1: 0.5375   auc_roc: 0.8033688453786352
depth: 9   F1: 0.560553633217993   auc_roc: 0.7906756886457433
depth: 10   F1: 0.5506756756756758   auc_roc: 0.7699851233323736
depth: 11   F1: 0.5322033898305084   auc_roc: 0.7589307995009118
depth: 12   F1: 0.5026178010471204   auc_roc: 0.7341923409156349
depth: 13   F1: 0.4972577696526508   auc_roc: 0.7139744697187831
depth: 14   F1: 0.4858757062146893   auc_roc: 0.69211056723294
depth: 15   F1: 0.50390625   auc_roc: 0.6961248680295613
depth: 16   F1: 0.4652087475149105   auc_roc: 0.672600537

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Здесь нужно заново попробовать разные гиперпараметры.

</div>


<div class="alert alert-info"> 
Благодарю, исправил
</div>

После балансировки резултат на лицо. При любом depth результат примерно 0.46382978723404256   auc_roc: 0.6646607159996161. Но нужно пробывать другую модель f1 мера маленькая

In [104]:
#На дереве решений это не отразилось

Случайный лес

In [105]:
for est in range(1, 200, 20):
    for depth in range(40, 100, 10):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth , class_weight = 'balanced') 
        model.fit(features_train_enc, target_train) 
        predicted_valid = model.predict(features_valid_enc)
        f1 = f1_score(target_valid, predicted_valid)
        probabilities_valid = model.predict_proba(features_valid_enc)
        probabilities_one_valid = probabilities_valid[:, 1]

        fpr, tpr, treshold = roc_curve(target_valid, probabilities_one_valid)
        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
        print('depth:', depth, ' ', 'F1:', f1, ' ', 'est:', est,  ' ', 'auc_roc:', auc_roc)

depth: 40   F1: 0.47537473233404715   est: 1   auc_roc: 0.6717679239850274
depth: 50   F1: 0.47537473233404715   est: 1   auc_roc: 0.6717679239850274
depth: 60   F1: 0.47537473233404715   est: 1   auc_roc: 0.6717679239850274
depth: 70   F1: 0.47537473233404715   est: 1   auc_roc: 0.6717679239850274
depth: 80   F1: 0.47537473233404715   est: 1   auc_roc: 0.6717679239850274
depth: 90   F1: 0.47537473233404715   est: 1   auc_roc: 0.6717679239850274
depth: 40   F1: 0.568421052631579   est: 21   auc_roc: 0.8352673001247719
depth: 50   F1: 0.568421052631579   est: 21   auc_roc: 0.8352673001247719
depth: 60   F1: 0.568421052631579   est: 21   auc_roc: 0.8352673001247719
depth: 70   F1: 0.568421052631579   est: 21   auc_roc: 0.8352673001247719
depth: 80   F1: 0.568421052631579   est: 21   auc_roc: 0.8352673001247719
depth: 90   F1: 0.568421052631579   est: 21   auc_roc: 0.8352673001247719
depth: 40   F1: 0.5767195767195767   est: 41   auc_roc: 0.8466527497840484
depth: 50   F1: 0.5767195767195

In [106]:
#А вот у случайного леса метрика улучшилась

После балансировки резултат на лицо. При любом depth и est результат примерно F1: 0.5745257452574525   est: 181   auc_roc: 0.8574983203762356 Но нужно пробывать другую модель f1 мера маленькая.

In [107]:
model = LogisticRegression(random_state=12345, solver='liblinear', class_weight = 'balanced')
model.fit(features_train_enc, target_train)
predicted_valid = model.predict(features_valid_enc)
f1 = f1_score(target_valid, predicted_valid)
probabilities_valid = model.predict_proba(features_valid_enc)
probabilities_one_valid = probabilities_valid[:, 1]
fpr, tpr, treshold = roc_curve(target_valid, probabilities_one_valid)
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
print('F1:', f1, ' ', 'auc_roc:', auc_roc) 

F1: 0.4933135215453195   auc_roc: 0.7806507342355312


In [108]:
#И у регрсесии метрика улучшилась

После балансировки резултат на лицо. F1: 0.4933135215453195   auc_roc: 0.7806507342355312. Но f1 мера маленькая.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Ок, зафиксировали.
</div>

2 стратегия метод SMOTE

In [109]:
#2 стратегия. 7237/1854 = 3,903452. Тоесть параметр repeat возбмем равным 4

In [110]:
#def upsample(features, target, repeat):
    #features_zeros = features[target == 0]
    #features_ones = features[target == 1]
    #target_zeros = target[target == 0]
    #target_ones = target[target == 1]

    #features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    #target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    #features_upsampled, target_upsampled = shuffle(
        #features_upsampled, target_upsampled, random_state=12345)
    
    #return features_upsampled, target_upsampled

#features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично, что используешь рукописную функцию, но здесь можно все оптимизировать, используя <a href="https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.SMOTE.html">SMOTE</a> для увеличения выборки и <a href="https://imbalanced-learn.org/stable/references/generated/imblearn.under_sampling.RandomUnderSampler.html">RandomUnderSampler</a>  для уменьшения. 
    
Апсемплинг делается очень просто:
    
`oversample = SMOTE(random_state=RANDOM_STATE)`

`features_train_up, target_train_up = oversample.fit_resample(features_train_enc, target_train)`
    
PS предварительно импотрировав `from imblearn.over_sampling import SMOTE`
    
PPS Чтобы импортировать imblearn его сначала требуется проинсталлировать. через `!pip install imblearn`

    
</div>

<div class="alert alert-info"> 
Благодарю, так сделаю
</div>

In [111]:
oversample = SMOTE(random_state=12345)
features_train_up, target_train_up = oversample.fit_resample(features_train_enc, target_train)
features_valid_up, target_valid_up = oversample.fit_resample(features_valid_enc, target_valid)
features_test_up, target_test_up = oversample.fit_resample(features_test_enc, target_test)

Решающее дерево

In [112]:
for depth in range(1, 50):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
    model.fit(features_train_up, target_train_up) 
    predicted_valid = model.predict(features_valid_up)
    f1 = f1_score(target_valid_up, predicted_valid)
    
    probabilities_valid = model.predict_proba(features_valid_up)
    probabilities_one_valid = probabilities_valid[:, 1]

    fpr, tpr, treshold = roc_curve(target_valid_up, probabilities_one_valid)
    auc_roc = roc_auc_score(target_valid_up, probabilities_one_valid)
    
    print('depth:', depth, ' ', 'F1:', f1, ' ', 'auc_roc:', auc_roc)

depth: 1   F1: 0.6501474926253688   auc_roc: 0.6727373068432672
depth: 2   F1: 0.6543463381245722   auc_roc: 0.7577731727165964
depth: 3   F1: 0.790750141003948   auc_roc: 0.8401958491099318
depth: 4   F1: 0.7970401691331924   auc_roc: 0.8593659391157309
depth: 5   F1: 0.7969029184038119   auc_roc: 0.8878990443888913
depth: 6   F1: 0.8241320432555492   auc_roc: 0.9008900681743978
depth: 7   F1: 0.8180764774044033   auc_roc: 0.8985388067774803
depth: 8   F1: 0.8011494252873562   auc_roc: 0.8902235038424241
depth: 9   F1: 0.8035190615835777   auc_roc: 0.8763668991126119
depth: 10   F1: 0.7890122735242548   auc_roc: 0.8481316608920662
depth: 11   F1: 0.7812685240071133   auc_roc: 0.8300427856478029
depth: 12   F1: 0.7979214780600461   auc_roc: 0.827720762734578
depth: 13   F1: 0.7829139847864247   auc_roc: 0.7947933570165051
depth: 14   F1: 0.7817437097717963   auc_roc: 0.799663391956493
depth: 15   F1: 0.7844276583381755   auc_roc: 0.791353571237129
depth: 16   F1: 0.7657342657342657   a

Метод SMOTE дает очень хороший результат. depth: 7 - 49 примерно  F1: 0.7596717467760844   auc_roc: 0.7737306843267109

Случайный лес

In [113]:
for est in range(1, 200, 20):
    for depth in range(1, 100, 10):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth) 
        model.fit(features_train_up, target_train_up) 
        predicted_valid = model.predict(features_valid_up)
        f1 = f1_score(target_valid_up, predicted_valid)
        
        probabilities_valid = model.predict_proba(features_valid_up)
        probabilities_one_valid = probabilities_valid[:, 1]

        fpr, tpr, treshold = roc_curve(target_valid_up, probabilities_one_valid)
        auc_roc = roc_auc_score(target_valid_up, probabilities_one_valid)
    
        print('depth:', depth, ' ', 'F1:', f1, ' ', 'est:', est,  ' ', 'auc_roc:', auc_roc)

depth: 1   F1: 0.6501474926253688   est: 1   auc_roc: 0.6727373068432672
depth: 11   F1: 0.7988636363636363   est: 1   auc_roc: 0.8570232056098903
depth: 21   F1: 0.77539175856065   est: 1   auc_roc: 0.787517238522677
depth: 31   F1: 0.7776484284051223   est: 1   auc_roc: 0.7891832229580573
depth: 41   F1: 0.7776484284051223   est: 1   auc_roc: 0.7891832229580573
depth: 51   F1: 0.7776484284051223   est: 1   auc_roc: 0.7891832229580573
depth: 61   F1: 0.7776484284051223   est: 1   auc_roc: 0.7891832229580573
depth: 71   F1: 0.7776484284051223   est: 1   auc_roc: 0.7891832229580573
depth: 81   F1: 0.7776484284051223   est: 1   auc_roc: 0.7891832229580573
depth: 91   F1: 0.7776484284051223   est: 1   auc_roc: 0.7891832229580573
depth: 1   F1: 0.7624132407901761   est: 21   auc_roc: 0.8479428290182205
depth: 11   F1: 0.8490456911509543   est: 21   auc_roc: 0.9330042785647803
depth: 21   F1: 0.8566433566433567   est: 21   auc_roc: 0.9312012387370924
depth: 31   F1: 0.8511383537653239   est

In [118]:
#Очень незначительные изминения

Очень незначительные изминения относительно решающего дерева. Лучший резкдьтат на всем диапозоне примерно одинаковй depth: 51   F1: 0.8529411764705883   est: 121   auc_roc: 0.9350156425887755

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Мы внесли значительные изменения во вводные данные, логично, что нужно заново подобрать гиперпараметры.
       
</div>


<div class="alert alert-info"> 
Благодарю, исправил
</div>

In [119]:
#А вот у случайного леса метрика немнго даже ухудшилась

Регрессия

In [120]:
model = LogisticRegression(solver='liblinear', random_state=12345) 
model.fit(features_train_up, target_train_up) 
predicted_valid = model.predict(features_valid_up)
f1 = f1_score(target_valid_up, predicted_valid)

probabilities_valid = model.predict_proba(features_valid_up)
probabilities_one_valid = probabilities_valid[:, 1]

fpr, tpr, treshold = roc_curve(target_valid_up, probabilities_one_valid)
auc_roc = roc_auc_score(target_valid_up, probabilities_one_valid)
    
print('F1:', f1, ' ', 'auc_roc:', auc_roc) 

F1: 0.7728487439871727   auc_roc: 0.8442904550969987


И у регрсесии метрика тоже улучгилась, F1: 0.7728487439871727   auc_roc: 0.8442904550969987

3 стратегия метод RandomUnderSampler

In [121]:
'''
#def downsample(features, target, fraction):
    #features_zeros = features[target == 0]
    #features_ones = features[target == 1]
    #target_zeros = target[target == 0]
    #target_ones = target[target == 1]

    #features_downsampled = pd.concat(
        [#features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    #target_downsampled = pd.concat(
        #[target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    #features_downsampled, target_downsampled = shuffle(
        #features_downsampled, target_downsampled, random_state=12345)
   #return features_downsampled, target_downsampled

#features_downsampled, target_downsampled = downsampled(features_train, target_train, 0,25)'''

'\n#def downsample(features, target, fraction):\n    #features_zeros = features[target == 0]\n    #features_ones = features[target == 1]\n    #target_zeros = target[target == 0]\n    #target_ones = target[target == 1]\n\n    #features_downsampled = pd.concat(\n        [#features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])\n    #target_downsampled = pd.concat(\n        #[target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])\n    \n    #features_downsampled, target_downsampled = shuffle(\n        #features_downsampled, target_downsampled, random_state=12345)\n   #return features_downsampled, target_downsampled\n\n#features_downsampled, target_downsampled = downsampled(features_train, target_train, 0,25)'

In [122]:
downsample = RandomUnderSampler(random_state=12345)

features_train_down, target_train_down = downsample.fit_resample(features_train_enc, target_train)
features_valid_down, target_valid_down = downsample.fit_resample(features_valid_enc, target_valid)
features_test_down, target_test_down = downsample.fit_resample(features_test_enc, target_test)

Решающее дерево

In [123]:
for depth in range(1, 50):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
    model.fit(features_train_down, target_train_down) 
    predicted_valid = model.predict(features_valid_down)
    f1 = f1_score(target_valid_down, predicted_valid)
    
    probabilities_valid = model.predict_proba(features_valid_down)
    probabilities_one_valid = probabilities_valid[:, 1]

    fpr, tpr, treshold = roc_curve(target_valid_down, probabilities_one_valid)
    auc_roc = roc_auc_score(target_valid_down, probabilities_one_valid)
    
    print('depth:', depth, ' ', 'F1:', f1, ' ', 'auc_roc:', auc_roc)

depth: 1   F1: 0.6592427616926503   auc_roc: 0.6673913043478261
depth: 2   F1: 0.6911447084233261   auc_roc: 0.736030245746692
depth: 3   F1: 0.6911447084233261   auc_roc: 0.7794706994328923
depth: 4   F1: 0.7574468085106384   auc_roc: 0.8126465028355387
depth: 5   F1: 0.7676348547717842   auc_roc: 0.8370699432892248
depth: 6   F1: 0.7672413793103448   auc_roc: 0.8272778827977315
depth: 7   F1: 0.7645788336933045   auc_roc: 0.8231001890359168
depth: 8   F1: 0.772823779193206   auc_roc: 0.8248582230623819
depth: 9   F1: 0.7531380753138075   auc_roc: 0.8001890359168241
depth: 10   F1: 0.7531380753138075   auc_roc: 0.7757655954631381
depth: 11   F1: 0.7574468085106384   auc_roc: 0.7732136105860113
depth: 12   F1: 0.7366167023554604   auc_roc: 0.7642533081285444
depth: 13   F1: 0.7172995780590717   auc_roc: 0.7337618147448016
depth: 14   F1: 0.7161016949152542   auc_roc: 0.7191871455576558
depth: 15   F1: 0.7291666666666667   auc_roc: 0.7322211720226843
depth: 16   F1: 0.7106382978723403  

In [112]:
#Очень незначительные изминения

Результат чуть хуже чем при SMOTE. depth: 49   F1: 0.7157894736842104   auc_roc: 0.7065217391304347

In [124]:
for est in range(1, 200, 20):
    for depth in range(1, 100, 10):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth) 
        model.fit(features_train_down, target_train_down) 
        predicted_valid = model.predict(features_valid_down)
        f1 = f1_score(target_valid_down, predicted_valid)
        
        probabilities_valid = model.predict_proba(features_valid_down)
        probabilities_one_valid = probabilities_valid[:, 1]
    
        fpr, tpr, treshold = roc_curve(target_valid_down, probabilities_one_valid)
        auc_roc = roc_auc_score(target_valid_down, probabilities_one_valid)
    
        print('depth:', depth, ' ', 'F1:', f1, ' ', 'est:', est,  ' ', 'auc_roc:', auc_roc)

depth: 1   F1: 0.6921529175050303   est: 1   auc_roc: 0.667391304347826
depth: 11   F1: 0.7028199566160521   est: 1   auc_roc: 0.7193289224952741
depth: 21   F1: 0.6709129511677283   est: 1   auc_roc: 0.6630434782608695
depth: 31   F1: 0.6709129511677283   est: 1   auc_roc: 0.6630434782608695
depth: 41   F1: 0.6709129511677283   est: 1   auc_roc: 0.6630434782608695
depth: 51   F1: 0.6709129511677283   est: 1   auc_roc: 0.6630434782608695
depth: 61   F1: 0.6709129511677283   est: 1   auc_roc: 0.6630434782608695
depth: 71   F1: 0.6709129511677283   est: 1   auc_roc: 0.6630434782608695
depth: 81   F1: 0.6709129511677283   est: 1   auc_roc: 0.6630434782608695
depth: 91   F1: 0.6709129511677283   est: 1   auc_roc: 0.6630434782608695
depth: 1   F1: 0.7543859649122807   est: 21   auc_roc: 0.8205103969754253
depth: 11   F1: 0.7789934354485777   est: 21   auc_roc: 0.8616635160680528
depth: 21   F1: 0.7833698030634574   est: 21   auc_roc: 0.8510964083175804
depth: 31   F1: 0.7868131868131868   e

У случайного леса метрика немнго eще ухудшилась при SMOTE. depth: 91   F1: 0.7895878524945769   est: 181   auc_roc: 0.8672873345935728

Регрессия

In [125]:
model = LogisticRegression(solver='liblinear', random_state=12345) 
model.fit(features_train_down, target_train_down) 
predicted_valid = model.predict(features_valid_down)
f1 = f1_score(target_valid_down, predicted_valid)

probabilities_valid = model.predict_proba(features_valid_down)
probabilities_one_valid = probabilities_valid[:, 1]
    
fpr, tpr, treshold = roc_curve(target_valid_down, probabilities_one_valid)
auc_roc = roc_auc_score(target_valid_down, probabilities_one_valid)
    
print('F1:', f1, ' ', 'auc_roc:', auc_roc) 

F1: 0.7336244541484717   auc_roc: 0.7900756143667296


И у регрсесии метрика тоже ухудшилась, и хуже чем у дерева и леса

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Везде, где возможно, нужно подбирать гиперпараметры.
       
</div>


<div class="alert alert-info"> 
Благодарю, исправил
</div>

Очень незначительные изминения

Вывод: 
Видно что самая лучшая оказалась первая стратегия где баланс классов устанавливается с помощью аргумента class_weight = 'balanced'. Самый лучший результат у модели случайного леса, где f1 метрика равна 0.6208. Что касается стратегии upsampling и downsampling, то следует отметить что нету точных показателей которые говорят насколько нужно увеличить или уменьшить выборку чтобы достичь наилучшего результата. Кто-то считает, что нужно выравнивать 50 на 50, есть те кто не использует upsampling/downsampling, кто-то пробует различные коэффициенты, это все эмпирические выводы.

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Как мы увидели, на линейных моделях дисбаланс классов становится серьезной проблемой для обучения. При огромном разбросе регрессии сложно подобрать коэффициенты к весам, не «перепрыгнув» оптимального состояния.
    
По балансу классов, в идеале нужно стремиться к 50/50.
       
</div>


<div class="alert alert-info"> 
Благодарю, принял. Как и писал выше я выводы оставою прежние чтобы не не терять нить
</div>

## Тестирование модели

Вначале сохраним лучшуу модель, случайный лес метод SMOKE при n_estimators=191, max_depth=91

In [126]:
model = RandomForestClassifier(random_state=12345, n_estimators=181, max_depth=91) 
model.fit(features_train_up, target_train_up) 
dump(model, 'model_RandomForestClassifier_SMOKE.joblib')

['model_RandomForestClassifier_SMOKE.joblib']

In [127]:
model = load('model_RandomForestClassifier_SMOKE.joblib')

test_predictions = model.predict(features_test_up)
test_f1 = f1_score(target_test_up, test_predictions)

probabilities_test = model.predict_proba(features_test_up)
probabilities_one_test = probabilities_test[:, 1]
fpr, tpr, treshold = roc_curve(target_test_up, probabilities_one_test)
auc_roc = roc_auc_score(target_test_up, probabilities_one_test)

print('depth:', depth, ' ', 'F1:', test_f1, ' ', 'est:', est,  ' ', 'auc_roc:', auc_roc)

depth: 91   F1: 0.8612385321100918   est: 181   auc_roc: 0.9411451679029768


Результат намного больше 0,59, depth: 91   F1: 0.8612385321100918   est: 181   auc_roc: 0.9411451679029768, но попробую обьединить выборки

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Хорошая идея.
</div>

In [128]:
NEW_features_train_up = pd.concat([features_train_up, features_valid_up])
NEW_target_train_up = pd.concat([target_train_up, target_valid_up])

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера #2 <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Мы здесь можем обединить трейн и валидационную, обучить модель на этой объединенной выборке заново, а затем посчитать метрику на тесте. Никак иначе.  В твоем случае у нас идет утечка целевого признака. 

</div>



<div class="alert alert-info"> 
Благодарю, исправил
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #3 <a class="tocSkip"> </h2>

👍 Отлично! Круто, что решил доделать.
</div>


In [134]:
model = RandomForestClassifier(random_state=12345, n_estimators=181, max_depth=91) 
model.fit(NEW_features_train_up, NEW_target_train_up) 
predicted_test = model.predict(features_test_up)
f1 = f1_score(target_test_up, predicted_test)


probabilities_test = model.predict_proba(features_test_up)
probabilities_one_test = probabilities_test[:, 1]
fpr, tpr, treshold = roc_curve(target_test_up, probabilities_one_test)
auc_roc = roc_auc_score(target_test_up, probabilities_one_test)
    
print('depth:', depth, ' ', 'F1:', f1, ' ', 'est:', est,  ' ', 'auc_roc:', auc_roc)

depth: 91   F1: 0.8649885583524027   est: 181   auc_roc: 0.9443815430674471


Результат не сильно поменялся 

<div class="alert alert-info"> 
Я правильно понял, что модель обучает при лучших гиперпараметрах которые нашои в преыдуший раз. Тоесть не нудно снова для этой выборки искать лучгий медот балансировки, лучшую модель и гиперпараметры.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Да, с тестом уже параметры подбирать нельзя. Иначе будет качество лучше, чем мы можем гарантировать заказчику. 
</div>


In [135]:
#Метрика улучшилась и стала равной 0,6086

In [136]:
#Иследуем метрику для нашей лучшей модели

In [137]:
#model = load('model_RandomForestClassifier.joblib')

#probabilities_valid = model.predict_proba(features_valid)
#probabilities_one_valid = probabilities_valid[:, 1]

#fpr, tpr, treshold = roc_curve(target_valid, probabilities_one_valid)
#auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
#print(auc_roc)

In [138]:
#AUC-ROC показывает как сильно наша модель отличается от случайной, это площадь под ROC-кривой. Это хороший результат. 

<div class="alert alert-info"> 
Результат больше 0,59 значит задача выполнена, но я так и неразоьрался почему выводит depth = 91, est = 191 хотя я указывал другие параметры
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Да, результат неплохой!
</div>

Вывод: Основной проблемой для адекватной работы модели был дисбаланс классов

Мы применили несколько стратегий борьбы с дисбалансом классов.

Стратегия class_weight = 'balanced'  дала следующие результаты:
• решающее дерево: F1-мера = 0.5601, при max_depth = 7
• случайный лес: максимальная F1-мера: 0.6208, при n_estimators = 200, при max_depth = 7 
• логистическая регрессия: F1-мера = 0.4962

Стратегия upsampling дала следующие результаты :
• решающее дерево: F1-мера = 0.5676, при max_depth = 7
• случайный лес: максимальная F1-мера: 0.6027, при n_estimators = 200, при max_depth = 7 
• логистическая регрессия: F1-мера = 0.4832

Стратегия downsampling дала следующие результаты :
• решающее дерево: F1-мера = 0.5681, при max_depth = 7
• случайный лес: максимальная F1-мера: 0.5837, при n_estimators = 200, при max_depth = 7 
• логистическая регрессия: F1-мера = 0.4892

В качестве лучшей модели была выбрана модель  случайного леса при значении гиперпараметра max_depth = 5, n_estimators = 200, указании атрибута class_weight='balanced' 

На валидационной выборке модель следующие результаты:
• F1-мера: 0.6208
На тестовой выборке модель следующие результаты:
• F1-мера: 0.60869
• AUC-ROC: 0.86463 

Проверку на тестовой выборке модель прошла. Требование касательно значения F1-меры (нужно довести метрику до 0.59) выполнено.

P.S. Для более лучшей работы стратегий по уменьшению дисблаоанса классов, следовало бы немного поиграть порогом и посмтреть как он повлияет на тосность и полноту и как следствие на f1 меру, но это более шлубокая работа и повидимому немного выодит за рамки проекта.

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Я предполагаю, что у нас достаточно большая выборка, как мы видели, многие признаки хорошо распределены, и оптимально будет все-таки доведение баланса до 50/50. Но всегда можно потестить.
       
</div>


<div class="alert alert-info"> 
Благодарю, вывод осталю. Тк логика таже, а цифры поменять всегда можно
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту**
    
Спасибо за твой проект! Мне было интересно следить за ходом мыслей и твоими решениями. Виден интерес к проекту. Есть несколько важных моментов, которые необходимо улучшить. Их ты найдешь в разноцветных комментариях.
    
**Отмечу положительные моменты проекта🚀**
    
* Выдержана структура проекта, хорошее оформление, аргументация шагов, выводы.
* Верные решения по предобработке.
* Хорошее исследование методов борьбы с дасбалансом.
    
**На что стоит обратить внимание🤔**
    
* До и после борьбы с дисбалансом в моделях нужно подбирать гиперпараметры, в «лесе» − и количество эстиматоров, и глубину.
* Оптимальнее использовать OHE из Sklearn. Рекомендую реализовать в проекте. Кодирование необходимо будет сделать после разбиения на выборки.
* Везде смотрим не только на F-меру, но и на ROC-AUC. Ты применил только к результату на тесте.
* Рекомендую также использовать SMOTE для ресемплирования «вверх». Это удобнее, чем рукописная функция.
* Эти и остальные замечания я также написал по ходу проекта.
    
**Желаю удачи и жду твой проект на повторное ревью!**😉
    
    
</div>


<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий ко второй проверке**

Павел, привет еще раз! Как понял, дошли до OHE. Я попробовал показать по коду и даже немного проверил работоспособность, все работает и вроде бы тебе останется доделать по аналогии. 
    
Описал подробно по двум параметрам, которые нам нужны будут примерно в 100% использования энкодера (по крайней мере в проектах Практикума). Также обрати внимание, что я ставлю определенные версии в начале проекта и OHE, и sklearn. В этих версиях проблем с исполнением точно быть не должно. Если где-то конспектируешь решения по коду, то на будущее запиши этот момент. В сборных проектах точно нужно будет использовать OHE и нужно будет устанавливать версии.
    
Дальше, надеюсь, с доработкой должно быть попроще. Там нужно почаще переиспользовать код с перебором гиперпараметров, а в лесе написать двойной цикл, чтобы и эстиматоры подобрать, и глубину. При апсемнлинге и даунсемплинге лучшие модели могут быть с разными параметрами, это надо учитывать.
    
Сразу выводим метрику и гиперпараметры, и сразу применяем кроме F1 еще ROC-AUC. Не настаиваю, но всегда красиво, когда в проекте есть хотя бы один график. Здесь будет смотреться график ROC-AUC. 
    
**Желаю удачи и жду твой проект на повторное ревью!**😉
</div>




<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий ко второй "+" проверке**

Чтобы OHE заработал, нужно применить верные импорты. Для этого: вверху ноутбука найти кнопку Kernel, затем в меню Restart&Ran All и нажать.
    
Я протестил, все работает. (дальше код падает уже из-за того, что мы недоприменили пока OHE, не объединили выборки ну и так далее).

</div>

<div class="alert alert-info"> 
Благодарю, вроде все исправоенно
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий к третьей проверке**

Спасибо за доработку. С главными сложностями мы справились.
    
Я взял на себя ответственность и обнулил индексы при объединении таблиц. Посмотри там красный комментарий, удали, наверное, двойную проверку на пропуски, заполнение нулями − всю эту часть. По всему проекту вопросов нет, метрики достигнуты шикарные.
    
В самом конце ты неверно объединяешь выборки - нужно объединить трейн и валидацинную - на них обучить модель с лучшими гиперпараметрами и еще раз померить на тесте. Вот так возможно сделать. Нужно либо переделать (если время терпит) либо снести этот блок. В остальном, повторюсь, проект готов. 
    
PS Я сначала загрустил, когда увидел результаты f1 (а потом как понял). Хороший проект получается :) И по всем неожиданным трудностям проехались. 
    
Осталось чуть-чуть. Жду на ревью!
</div>

<div class="alert alert-info"> 
Благодарю, Удалил комментарий и проверку на нули с заполнением. 
    
По поводу обьединения - это опыт предыдущего проекта, тк там говорили валидационную с тестовой обьелинять и тем самым большая тестовая будет. И на этой большей тестовой выборке проверить лучшую модель обученную на трейне. 
    
А так если я обьединяю трейн и валид, то у меня будет просто большой трейн и маленькая тестовая. Время терпит, но я пока не понял смысл этой процедуры, попробую чисто из интереса. 
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #3 <a class="tocSkip"> </h2>

👍 Да, тут история только в том, что чем больше трейн, тем лучше обобщение модели и тем меньше переобучение. По ходу работы мы не можем работать без валидационной выборки (почти, если не считать кросс-валидацию), потом же нам валидационная как отдельная единица уже не нужна, можно объединить и посмотреть на качество модели. 
</div>


<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий к финальной проверке**

Спасибо за оперативную доработку. Критичных комментариев нет. Интересный проект вышел! 

Желаю удачи в дальнейшей учебе!

</div>


## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [x]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*